### PR TITLE
CDR-1532 Update Postgres v15 migration

### PR DESCRIPTION
**Pull request by alexlehn**
# Changes

Update pg migration for v15.

# Pre-Merge checklist

- [x] New code is tested
- [x] Present and new tests pass
- [x] Documentation is updated
- [x] The build is working without errors
- [x] No new Sonar issues introduced
- [ ] Changelog is updated
- [ ] Code has been reviewed

### CODE REVIEW

File: jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql [hunk 1]
```diff
@@ -16,137 +16,275 @@
  * limitations under the License.
  */
 
-ALTER TABLE comp_data
+
+-- Ensure clean state
+DROP FUNCTION IF EXISTS ext.mig_calc_nums(ext.mig_num_type[]);
+DROP FUNCTION IF EXISTS ext.mig_retrieve_nums_batch(text, text, text, text, text, text, integer);
+DROP PROCEDURE IF EXISTS ext.mig_num_columns(regclass, text, text, integer);
+DROP TYPE IF EXISTS ext.mig_num_type CASCADE;
```
Comment (HolgerReiseVSys) [inline jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql:24]: Move all the functions etc. to schema pg_temp.
Dropping mig_num_type will also drop the functions (but not the procedure).
Also, the CREATE OR REPLACE on the procedure should be sufficient.

File: jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql [hunk 2]
```diff
@@ -16,137 +16,275 @@
...
-DROP INDEX IF EXISTS mig_ehr_status_data_history_num_idx;
-DROP INDEX IF EXISTS mig_ehr_status_data_num_idx;
-DROP INDEX IF EXISTS mig_comp_data_history_num_idx;
-DROP INDEX IF EXISTS mig_comp_data_num_idx;
+--
+-- Migrate
+--
+
+--may be executed in parallel
+CALL ext.mig_num_columns('ehr_folder_data_history'::regclass, 'ehr_id', 'mig_ehr_folder_data_history_num_idx',1000);
```
Comment (HolgerReiseVSys) [inline jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql:274]: Reduce to maybe 100. Some systems tend to have extensive folder histories.

File: jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql [hunk 3]
```diff
@@ -16,137 +16,275 @@
...
+    return rs;
+END;$function$;
+
+--
+-- Migration procedure
+--
+CREATE OR REPLACE PROCEDURE ext.mig_num_columns(
+    rel_name regclass,
+    id_exp text,
+    mig_idx text,
```
Comment (HolgerReiseVSys) [inline jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql:144]: mig_idx is not needed

File: jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql [hunk 4]
```diff
@@ -16,137 +16,275 @@
...
+    batch_size integer DEFAULT 1000
+) LANGUAGE plpgsql
+AS $procedure$
+DECLARE
+    rs ext.mig_num_type[];
+    cols text;
+    ehr_id_exp text;
+    folders_idx_exp text;
```
Comment (HolgerReiseVSys) [inline jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql:152]: folders_idx_exp is not needed

File: jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql [hunk 5]
```diff
@@ -16,137 +16,275 @@
...
+
+    ehr_id_exp=COALESCE(substring(cols FROM '(?i)ehr_id'),'NULL');
+    folders_idx_exp=COALESCE(substring(cols FROM '(?i)ehr_folders_idx'),'NULL');
+    version_exp=COALESCE(substring(cols FROM '(?i)sys_version'),'1::int');
+
+    last = clock_timestamp();
+
+    WHILE updated_cnt <> 0 LOOP
+            COMMIT;
+            SET TRANSACTION ISOLATION LEVEL SERIALIZABLE READ ONLY DEFERRABLE;
```
Comment (HolgerReiseVSys) [inline jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql:179]: Separate transactions are only needed for pre-flight.
The initial and intermediate commit are not needed. 
Also, catching exceptions is not needed.

**Review by HolgerReiseVSys — CHANGES_REQUESTED** [review]

File: jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql (hunk 5 above)
Comment (alexlehn) [reply jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql:179]: Intension was to be able to use this script as a pre-flight.

Comment (HolgerReiseVSys) [reply jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql]: Not sure about the impact, but it needs 3x as many transactions + transaction isolation

File: jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql (hunk 3 above)
Comment (vmueller-vg) [reply jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql]: @alexlehn there might have been a misunderstanding:
We do not need the procedure argument, but the indexes still need to be created

File: jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql [hunk 6]
```diff
@@ -16,137 +16,240 @@
...
-DROP INDEX IF EXISTS mig_comp_data_history_num_idx;
-DROP INDEX IF EXISTS mig_comp_data_num_idx;
-
-DROP INDEX IF EXISTS comp_data_idx;
-DROP INDEX IF EXISTS comp_data_leaf_idx;
-DROP INDEX IF EXISTS ehr_status_data_idx;
+--
+-- Migrate
+--
+
```
Comment (vmueller-vg) [inline jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql:237]: The migration indexes are missing

Comment (alexlehn) [reply jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql:237]: now I'm confused -> https://github.com/ehrbase/ehrbase/pull/1389#discussion_r1718247751

File: jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql (hunk 5 above)
Comment (alexlehn) [reply jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql]: it's removed in this commit -> https://github.com/ehrbase/ehrbase/pull/1389/commits/781dae2cc444cf78bd0c7e43ea7daa433b6bf6a6

File: jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql (hunk 3 above)
Comment (alexlehn) [reply jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql]: and they are back again https://github.com/ehrbase/ehrbase/pull/1389/commits/56ee6c1d550d9dc2b3b881f5630e48de5ddd3df1

File: jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql (hunk 4 above)
Comment (alexlehn) [reply jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql]: is now removed

File: jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql (hunk 2 above)
Comment (alexlehn) [reply jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql]: done

**Review by HolgerReiseVSys — APPROVED** [review]

**Pull request merged**